types/query/filtered_pagination.go

```diff
@@ -114,3 +119,137 @@ func FilteredPaginate(
 
 	return res, nil
 }
+
+// GenericFilteredPaginate does pagination of all the results in the PrefixStore based on the
+// provided PageRequest. `onResult` should be used to filter or transform the results.
+// `c` is a constructor function that needs to return a new instance of the type T (this is to
+// workaround some generic pitfalls in which we can't instantiate a T struct inside the function).
+// If key is provided, the pagination uses the optimized querying.
+// If offset is used, the pagination uses lazy filtering i.e., searches through all the records.
+// The resulting slice (of type F) can be of a different type than the one being iterated through
+// (type T), so it's possible to do any necessary transformation inside the onResult function.
+func GenericFilteredPaginate[T codec.ProtoMarshaler, F codec.ProtoMarshaler](
```

Comment:
Why do we need `T` and `F`, should just `[T codec.ProtoMarshaler]` suffice?

Reply:
Because in some cases we might want to do a transformation, like here: https://github.com/cosmos/cosmos-sdk/blob/40cd7b62b1167f7963a982a8017487207d6fe696/x/authz/keeper/grpc_query.go#L106